### PR TITLE
OIDC session expires with ID token expiry and refresh token is never used.

### PR DESCRIPTION
**Pull request by Alankarsharma**
### Description
In OIDC flow ID token expiry time was used to set cookie expiry time. Because of this cookie was getting expired at same time as ID token and alternate flow in isValidCookie function to fetch new ID token using refresh token is not getting triggered as cookie itself is expired.

### Category
Bug fix

### What is the old behavior before changes and new behavior after changes?
This code was correct in version 2.11.1, issue is only in 2.13.0 version. I haven't verified 2.12.0 version.

### Issues Resolved
[List any issues this PR will resolve (Is this a backport? If so, please add backport PR # and/or commits #)]

### Testing
[Please provide details of testing done: unit testing, integration testing and manual testing]

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by derek-ho** [comment]
Hi @Alankarsharma thanks for the PR! Just providing some initial feedback here - I was the person who made the PR that changed the OIDC cookie expiry - initial PR is [here](https://github.com/opensearch-project/security-dashboards-plugin/pull/1773). I am not an OIDC expert, so maybe I got some things wrong, but we can discuss here and get to the bottom of this (I still have seen some complaints about auto logouts in the public slack, and I am actually trying to replicate and dive deep on this myself - see conversation here: https://opensearch.slack.com/archives/C051Y637FKK/p1712521239951639).

Some basic background is that there is actually two checks that go into seeing if the cookie is still valid - the place you modified in this PR, but also [here](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/session/security_cookie.ts#L71). In my initial PR in which I consolidated the expiryTime variables, and used the expiry from the token response is because from my understanding for external IDP login setups we should not be blindly giving a session length of TTL, and should instead rely on the timeout from the credentials given by the IDP. Is that not correct/can you explain a little more the reasoning behind why you think this PR would fix the issues?

File: server/auth/types/openid/openid_auth.ts [hunk 1]
```diff
@@ -272,7 +272,7 @@ export class OpenIdAuthentication extends AuthenticationType {
       return false;
     }
 
-    if (cookie.expiryTime > Date.now()) {
+    if (cookie.credentials.expiryTime > Date.now()) {
```
Comment (DarshitChanpura) [inline server/auth/types/openid/openid_auth.ts:275]: Can you add a little more explanation on how setting expiryTime directly from credentials would resolve the issue at hand?

Comment (Alankarsharma) [reply server/auth/types/openid/openid_auth.ts:275]: OIDC response has two JWT token, ID Token and Refresh token. You are taking expiry time of ID token and setting it   as expiry of cookie, ID token are short lived token and once they are expired we can get new token using Refresh token, code for this is present in isValidCookie. 
In login flow, I have set ID token expiry value in 'cookie.credentials.expiry' and checking if ID token is still valid or not. If it is valid then true is returned, if it is not valid a new ID token will be fetched using Refresh token and value of this token are updated in cookie again.
User is logged-out only if both ID token and refresh token are expired or cookie is invalidated.

**Review by derek-ho — COMMENTED** [review]
Thanks @Alankarsharma I took another look through the whole flow - I think this PR probably also needs to introduce a check before this line: https://github.com/opensearch-project/security-dashboards-plugin/blob/6425893f33cc9047753340403f37cadf9cb6d373/server/session/security_cookie.ts#L71 to switch the expirytime variable for OIDC. Additionally, we may want to change the keepAlive setting to add the TTL onto the expiry cookie instead of being a no-op here: https://github.com/opensearch-project/security-dashboards-plugin/blob/6425893f33cc9047753340403f37cadf9cb6d373/server/auth/types/openid/openid_auth.ts#L258. Can you also run `yarn lint:es --fix` to address the lint failures on this PR and fix any test cases that are failing? @cwperks do you agree with this assessment?

File: server/auth/types/openid/openid_auth.ts [hunk 2]
```diff
@@ -296,8 +296,9 @@ export class OpenIdAuthentication extends AuthenticationType {
           cookie.credentials = {
             authHeaderValueExtra: true,
             refresh_token: refreshTokenResponse.refreshToken,
+            expiryTime: getExpirationDate(refreshTokenResponse),
           };
-          cookie.expiryTime = getExpirationDate(refreshTokenResponse);
+          cookie.expiryTime = Date.now() + this.config.session.ttl;
```
Comment (RyanL1997) [inline server/auth/types/openid/openid_auth.ts:301]: If there is a session `ttl` that is longer than the original `expiryTime`, will this be also added into the updated `expiryTime`?

**Review by RyanL1997 — COMMENTED** [review]
Hi @Alankarsharma, thanks for taking this on.

File: server/auth/types/openid/openid_auth.ts (hunk 2 above)
Comment (Alankarsharma) [reply server/auth/types/openid/openid_auth.ts:301]: session `ttl` is only for cookie, if `expiryTime` is less than session `ttl` and refresh token is also invalidated or not present then `isCookieValid` function will return false and cookie will get invalidated.

**Comment by Alankarsharma** [comment]
@derek-ho I checked two code location that you mentioned. No code change is required there.

**Comment by Alankarsharma** [comment]
@derek-ho I will check test case for refresh token, will create a separate pull request for that.

**Comment by cwperks** [comment]
@Alankarsharma Thank you for this PR! I branched off of this and added a test to verify the refreshToken workflow with this fix included: https://github.com/opensearch-project/security-dashboards-plugin/pull/1990

I believe there's one more area where the expiryTime of the credentials needs to be updated instead of the cookie here: https://github.com/opensearch-project/security-dashboards-plugin/blob/9abc57d415e8eedfc96d1bb399cfe118500a4d21/server/auth/types/openid/openid_auth.ts#L296-L300

The changes in this PR will work for the first refresh, but would fail on subsequent refresh since it would be setting the expiry time of the cookie to the expiry time of the access token after first refresh.

**Comment by derek-ho** [comment]
@Alankarsharma will be closing this out since the change was merged with @cwperks PR. Thank you for your contribution! I hope to see more from you in the future :)

**Pull request closed without merging**